boa_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,345 @@
+//! Boa parser implementation.
+
+mod cursor;
+mod expression;
+mod statement;
+
+pub(crate) mod function;
+
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    error::ParseResult,
+    lexer::TokenKind,
+    parser::{
+        cursor::Cursor,
+        function::{FormalParameters, FunctionStatementList},
+    },
+    Error,
+};
+use boa_interner::Interner;
+use boa_macros::utf16;
+use rustc_hash::FxHashSet;
+use std::io::Read;
+
+use boa_ast::{
+    expression::Identifier,
+    function::FormalParameterList,
+    operations::{
+        contains, top_level_lexically_declared_names, top_level_var_declared_names, ContainsSymbol,
+    },
+    Position, StatementList,
+};
+
+/// Trait implemented by parsers.
+///
+/// This makes it possible to abstract over the underlying implementation of a parser.
+trait TokenParser<R>: Sized
+where
+    R: Read,
+{
+    /// Output type for the parser.
+    type Output; // = Node; waiting for https://github.com/rust-lang/rust/issues/29661
+
+    /// Parses the token stream using the current parser.
+    ///
+    /// This method needs to be provided by the implementor type.
+    fn parse(self, cursor: &mut Cursor<R>, interner: &mut Interner) -> ParseResult<Self::Output>;
+}
+
+/// Boolean representing if the parser should allow a `yield` keyword.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AllowYield(bool);
+
+impl From<bool> for AllowYield {
+    fn from(allow: bool) -> Self {
+        Self(allow)
+    }
+}
+
+/// Boolean representing if the parser should allow a `await` keyword.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AllowAwait(bool);
+
+impl From<bool> for AllowAwait {
+    fn from(allow: bool) -> Self {
+        Self(allow)
+    }
+}
+
+/// Boolean representing if the parser should allow a `in` keyword.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AllowIn(bool);
+
+impl From<bool> for AllowIn {
+    fn from(allow: bool) -> Self {
+        Self(allow)
+    }
+}
+
+/// Boolean representing if the parser should allow a `return` keyword.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AllowReturn(bool);
+
+impl From<bool> for AllowReturn {
+    fn from(allow: bool) -> Self {
+        Self(allow)
+    }
+}
+
+/// Boolean representing if the parser should allow a `default` keyword.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AllowDefault(bool);
+
+impl From<bool> for AllowDefault {
+    fn from(allow: bool) -> Self {
+        Self(allow)
+    }
+}
+
+#[derive(Debug)]
```

Comment:
We should probably add some documentation to the parser.

Reply:
I'll add some docs in the crate root.